dio/lib/src/transformers/util/transform_empty_to_null.dart

```diff
@@ -0,0 +1,48 @@
+import 'dart:async';
+import 'dart:typed_data';
+
+/// A [StreamTransformer] that replaces an empty stream of Uint8List with a default value
+/// - the utf8-encoded string "null".
+/// Feeding an empty stream to a JSON decoder will throw an exception, so this transformer
+/// is used to prevent that; the JSON decoder will instead return null.
+class DefaultNullIfEmptyStreamTransformer
+    extends StreamTransformerBase<Uint8List, Uint8List> {
+  const DefaultNullIfEmptyStreamTransformer();
+
+  @override
+  Stream<Uint8List> bind(Stream<Uint8List> stream) {
+    return Stream.eventTransformed(
+      stream,
+      (sink) => _DefaultIfEmptyStreamSink(sink),
+    );
+  }
+}
+
+class _DefaultIfEmptyStreamSink implements EventSink<Uint8List> {
+  _DefaultIfEmptyStreamSink(this._outputSink);
+
+  /// Hard-coded constant for replacement value, "null"
+  static final Uint8List _nullUtf8Value =
+      Uint8List.fromList(const [110, 117, 108, 108]);
```

Comment:
It's for adding the byte sequence for "null" to the response stream, in the case that the response stream is unexpectedly empty. 
So we return `null` instead of throwing an exception.

Reply:
For clarification: The response won't become the String `"null"`, it would become the value `null` after the json decoder decodes that byte sequence.

It has always been the behavior of Dio to return `null` when the `ResponseType` is `.json` and the response is empty.
Only in the recent release with the `FusedTransformer` there is now a corner case that if 
- the `ResponseType` is `.json` and
- the `Content-Type` header is `application/json` and
- the `Content-Length` header has a value of > 0 and
- the response is still empty

That an exception is thrown (instead of the old behavior of returning null)

I would also be okay with removing the behavior of returning null on empty responses when the `ResponseType` is `.json` but the body is empty completely, and always throw an exception in that case.
But that would be a breaking change for sure.

The behavior of the current release, where it mostly returns `null` on empty responses but throws an exception in certain corner cases (mostly responses that are invalid according to the HTTP spec, but still occur in the real world), is not ideal IMO.